---
title: "Comparison of proteomics and RNA-seq"
output: html_notebook
author: "Roy Novoselsky"
date: "April 2024"
---

```{r}
rm(list=ls()) 
```

# dependencies
```{r}
library(tidyr)
library(dplyr)
library(ggpubr)
library(patchwork)
library(ggrepel)

directory = getwd()
directory_split = unlist(strsplit(directory, '/'))
root_dir =  paste0(directory_split[-c(length(directory_split), length(directory_split)-1)], collapse = '/')
source(paste(root_dir,"/functions.R",sep=""))
dir_output = paste(directory,"output",sep="/")
if (!dir.exists(dir_output)) {dir.create(dir_output)}
```

# import data
```{r}
data = read.csv(paste(root_dir,"human_proteomics","output","processed_data.csv",sep="/"))
rownames(data) = data$gene
prot_raw = data[1:8]

prot_med = read.csv(paste(root_dir,"human_proteomics","output","processed_data.csv",sep="/"))
colnames(prot_med)[colnames(prot_med)=="a_b_ratio"] = "log2_fc"

rna_raw = read.csv(paste(root_dir,"human_rna","output","export_data.csv",sep="/"),row.names=1)
rna_med = read.csv(paste(root_dir,"human_rna","output","medians.csv",sep="/"),row.names=1)
```
# merge median values of apicome
```{r}
columns = "gene|expression_mean|log2_fc|qval$"
rna_prot = merge(rna_med[,grepl(columns,colnames(rna_med))],prot_med[,grepl(columns,colnames(prot_med))],by="gene",suffixes=c("_rna","_prot"))
```
# Check zonation correlation (tip/base)
```{r}
rna_med$tip_rna = (rna_med$apical_tip + rna_med$basal_tip)/2
rna_med$base_rna = (rna_med$apical_base + rna_med$basal_base)/2

rna_med$ratio_rna = log2((rna_med$tip_rna) / (rna_med$base_rna))
rna_med$expression_mean = apply(rna_med[,c("tip_rna","base_rna")],1,mean)
rna_med$gene = rownames(rna_med)

pn = min(c(min(prot_med$apical_tip[prot_med$apical_tip>0]),
           min(prot_med$apical_base[prot_med$apical_base >0])))
# pn = 1e-6
prot_med$ratio_protein = log2((prot_med$apical_tip+pn)/(prot_med$apical_base+pn))
# prot_med$ratio_protein = prot_med$log2_fc

merged = merge(rna_med[c("expression_mean","gene","ratio_rna")],
               prot_med[,c("ratio_protein","expression_mean","gene")],
               by="gene",suffixes = c("_rna","_protein"))
```


```{r}
exp_thresh = 1e-4
plot = merged[(merged$expression_mean_protein > exp_thresh) &
                (merged$expression_mean_rna > exp_thresh),]
ggplot(plot, aes(ratio_rna, ratio_protein)) +
  geom_point() +
  stat_cor(method="spearman",digits=2) +
  theme_bw(base_size=14) +
  xlab("log2(tip/base) - mRNA") +  ylab("log2(tip/base) - Protein") +
  geom_hline(yintercept=0, color="black",alpha = 0.2,linetype="dashed") +
  geom_vline(xintercept=0, color="black",alpha = 0.2,linetype="dashed") +
  theme(legend.position="none",panel.grid=element_blank(),axis.text=element_text(color="black")) +
  ggtitle("")

ggsave(paste(dir_output,"/mRNA_protein_zonation_corr.pdf",sep=""),width=6,height=6,units="in")
```

# non-normilized correlation mRNA/protein
## plot spearman correlation
```{r}
EXPRESS_THRESH = 1e-5
plot = rna_prot[rna_prot$expression_mean_rna > EXPRESS_THRESH & rna_prot$expression_mean_prot > EXPRESS_THRESH,]
plot$signif = F
plot$signif[grepl("CDH1$|MYH14|SLC5A1",plot$gene)] = T
ggplot(plot, aes(log2_fc_rna, log2_fc_prot)) +
  geom_point(aes(color=signif,size=signif,fill=signif),shape=21,fill="black") +
  stat_cor(method="spearman",digits=2,label.y=10) +
  geom_text_repel(data=plot[plot$signif == T,],aes(log2_fc_rna, log2_fc_prot,label=gene,color=signif),max.overlaps = 30, size=3) +
  geom_text_repel(data=plot[plot$signif == F,],aes(log2_fc_rna, log2_fc_prot,label=gene),max.overlaps = 5, size=3)+
  theme_bw(base_size=18) +
  scale_color_manual(values = c("black","red")) +
  scale_size_manual(values=c(1,3))+
  xlab("log2(apical/basal) - mRNA") +  ylab("log2(apical/basal) - Protein") +
  geom_hline(yintercept=0, color="black",alpha = 0.2,linetype="dashed") +
  geom_vline(xintercept=0, color="black",alpha = 0.2,linetype="dashed") +
  theme(legend.position="none",panel.grid=element_blank(),axis.text=element_text(color="black"))+
  ggtitle("all genes")

ggsave(paste(dir_output,"/RNA_protein_scatter_all.pdf",sep=""),width=6,height=6,units="in")
# shell.exec(paste(dir_output,"/RNA_protein_scatter_all.pdf",sep=""))
```

# show number of coordant/discordant genes
## table
```{r}
EXPRESS_THRESH = 0
plot = rna_prot[rna_prot$expression_mean_rna >= EXPRESS_THRESH & rna_prot$expression_mean_prot >= EXPRESS_THRESH,]

thresh_prot = 0
thresh_rna = 0


count_rna_apical = sum(plot$log2_fc_rna > thresh_rna)
count_rna_basal = sum(plot$log2_fc_rna < -thresh_rna)
count_prot_apical = sum(plot$log2_fc_prot > thresh_prot)
count_prot_basal = sum(plot$log2_fc_prot < -thresh_prot)

quarters = data.frame(
  q = c(1,2,3,4),
  type = c("Coordant","discordant","Coordant","discordant"),
  mRNA = c("Apical\nmRNA","Basal\nmRNA","Basal\nmRNA","Apical\nmRNA"),
  Protein = c("Apical\nprotein","Apical\nprotein","Basal\nprotein","Basal\nprotein"),
  count_rna = c(count_rna_apical,count_rna_basal,count_rna_basal,count_rna_apical),
  count_prot = c(count_prot_apical,count_prot_apical,count_prot_basal,count_prot_basal),
  count_both = c(sum(plot$log2_fc_prot > thresh_prot & plot$log2_fc_rna > thresh_rna),
            sum(plot$log2_fc_prot > thresh_prot & plot$log2_fc_rna < -thresh_rna),
            sum(plot$log2_fc_prot < -thresh_prot & plot$log2_fc_rna < -thresh_rna),
            sum(plot$log2_fc_prot < -thresh_prot & plot$log2_fc_rna > thresh_rna)
  ),
  population_size = rep(nrow(plot),4),
  pval = c(NA,NA,NA,NA)
)

quarters$mRNA = factor(quarters$mRNA,levels=c("Basal\nmRNA","Apical\nmRNA"))
quarters$Protein = factor(quarters$Protein,levels=c("Basal\nprotein","Apical\nprotein"))


quarters$pval = phyper(q=quarters$count_both,
                       m=quarters$count_rna,
                       n=quarters$count_prot,
                       k=quarters$population_size,
                       lower.tail=F)

quarters$pval = phyper(q=quarters$count_both, 
  m=quarters$count_rna, 
  n=quarters$population_size - quarters$count_rna, 
  k=quarters$count_prot,
  lower.tail=F)

quarters$pval_sci = sapply(quarters$pval, function(p) {
  formatted_p = ifelse(p < 0.01, sprintf("%.2e", p), sprintf("%.1f", p))
  gsub("e(-?)0", "e\\1", formatted_p)
})

quarters$percent = round(quarters$count_both/sum(quarters$count_both)*100,1)
quarters$label = paste0(quarters$count_both," (",quarters$percent,"%)\n","p=", quarters$pval_sci)

colors = c("springgreen2","khaki1")

p1 = ggplot(quarters, aes(x = mRNA, y = Protein)) +
  # geom_tile(fill = "lightblue", color = "darkblue", size = 1.2) + 
  geom_tile(aes(fill=type), color = "darkblue", size = 1.2) + 
  scale_fill_manual(values=colors)+
  geom_text(aes(label = label), size = 6) + 
  scale_x_discrete(expand = c(0, 0)) +
  scale_y_discrete(expand = c(0, 0)) +
  xlab(NULL) + ylab(NULL) + 
  theme_minimal() + 
  coord_fixed(ratio = 0.7)+
  theme(axis.text = element_text(color="black",size=16),
        legend.position = "none",
        plot.margin=margin(0,0,0,0),
        panel.border = element_rect(color="darkblue", fill=NA, size=2))
# ggsave(paste(dir_output,"/hypergeometric.pdf",sep=""),width=5,height=4,units="in")
# shell.exec(paste(dir_output,"/hypergeometric.pdf",sep=""))

p2 = pie_chart(c("Coordant","Discordant"),
          c(sum(quarters$count_both[quarters$q %in% c(1,3)]),
            sum(quarters$count_both[quarters$q %in% c(2,4)])),
          colors = colors,
          thickness = 1.5) 
p2 = p2 + theme(legend.text = element_text(size=8),plot.margin=margin(0,0,0,0))

# p2 / p1 
 p2 / p1 + plot_layout(ncol = 1, heights = c(1.5, 1))

ggsave(paste(dir_output,"/piechart.pdf",sep=""),width=6,height=6,units="in")
shell.exec(paste(dir_output,"/piechart.pdf",sep=""))

```

# filter for enterocyte specific genes
## load scRNAseq data, from Harnik et al. 2024.
```{r}
dataset = read.csv(paste(directory,"input","signature_matrix_yotams.csv",sep="/"),row.names=1)

dataset = dataset[rowSums(dataset)>0,]
dataset$Best4 = NULL
dataset$Enterocytes = dataset$Enterocytes..mature.
dataset = matnorm(dataset)
dataset[,colnames(dataset)[grepl("Enterocytes.",colnames(dataset))]] = NULL
celltypes = colnames(dataset)

epi = dataset[,celltypes[grepl("Enterocytes|Fibroblasts",celltypes)]]
epi$Fibroblasts = NULL

other = dataset[,celltypes[!grepl("Enterocytes",celltypes)]]

epi_max = apply(epi,1,max)
other_max = apply(other,1,max)

epithelial_specific = data.frame(epi_max,other_max)
colnames(epithelial_specific) = c("enterocyte","other")

pn = min(epithelial_specific[epithelial_specific>0])
epithelial_specific$ratio = (epithelial_specific$enterocyte+pn)/(epithelial_specific$other+pn)

epithelial_specific$gene = rownames(epithelial_specific)
epithelial_specific = epithelial_specific[order(epithelial_specific$ratio,decreasing=T),]

# epithelial_specific[c("CPS1","NET1","MYH14","CDH1","DSP","HEPH","APOA1","MUC2"),]

```

# plot (epithelial specific normilized)
```{r}
thresh_ratio = 2
thresh_expr = 1e-5

epithelial_specific_genes = epithelial_specific$gene[epithelial_specific$ratio>thresh_ratio & epithelial_specific$enterocyte > thresh_expr ]

# merge
rna_epi = rna_raw[rownames(rna_raw) %in% epithelial_specific_genes,]
prot_epi = prot_raw[rownames(prot_raw) %in% epithelial_specific_genes,]

# normilize over enterocyte genes
rna_epi = matnorm(rna_epi)
prot_epi = matnorm(prot_epi)

vector = colnames(rna_epi)[colnames(rna_epi)!="gene"]
patients = unique(sapply(strsplit(vector, "_"), function(x) if(length(x) >= 3) x[3] else NA))

pn = min(rna_epi[,!grepl("gene",colnames(rna_epi))][rna_epi[,!grepl("gene",colnames(rna_epi))]>0])

# calculate geometric averege of ratios
for (p in patients){
  rna_epi[[paste0("ratio_",p)]] = (rna_epi[,paste0("apical_tip_",p)]+pn)/(rna_epi[,paste0("basal_tip_",p)]+pn)
}
rna_epi[,grepl("ratio_",colnames(rna_epi))] = log(rna_epi[,grepl("ratio_",colnames(rna_epi))],2)

rna_epi$a_b_ratio = apply(rna_epi[,grepl("ratio_",colnames(rna_epi))],1,mean)
rna_epi$gene = rownames(rna_epi)
pn = min(prot_epi[,!grepl("gene",colnames(prot_epi))][prot_epi[,!grepl("gene",colnames(prot_epi))]>0])

prot_epi$apical = apply(prot_epi[,grepl("apical_tip",colnames(prot_epi))],1,median) 
prot_epi$basal = apply(prot_epi[,grepl("basal_tip",colnames(prot_epi))],1,median) 
prot_epi$ratio = (prot_epi$apical+pn)/(prot_epi$basal+pn)
prot_epi$a_b_ratio = log(prot_epi$ratio,2)
prot_epi$gene = rownames(prot_epi)

plot = merge(prot_epi[,c("a_b_ratio","gene")],rna_epi[,c("a_b_ratio","gene")],by="gene",suffixes=c("_prot","_rna"))

ggplot(plot, aes(a_b_ratio_rna, a_b_ratio_prot)) +
  geom_point(color="black") +
  stat_cor(method="spearman",digits=2) + 
  geom_text_repel(aes(label=gene),max.overlaps = 3, size=3)+
  theme_bw(base_size=18) +
  xlab("log2(apical/basal) - mRNA") +  ylab("log2(apical/basal) - Protein") +
  geom_hline(yintercept=0, color="black",alpha = 0.2,linetype="dashed") +
  geom_vline(xintercept=0, color="black",alpha = 0.2,linetype="dashed") +
  theme(legend.position="none",panel.grid=element_blank(),axis.text=element_text(color="black")) +
  ggtitle(paste0("ratio: ",thresh_ratio,",  expression > ",thresh_expr))

ggsave(paste(dir_output,"/RNA_protein_scatter_normilized_enterocyte.pdf",sep=""),width=6,height=6,units="in")
```

# export
```{r}
write.csv(rna_prot, paste(dir_output,"/rna_prot.csv",sep=""))
write.csv(epithelial_specific, paste(dir_output,"/epithelial_specific_genes.csv",sep=""))
```
